### PR TITLE
Query: Use GetDiscriminatorValue for TPT and TPC

### LINKED ISSUE #28054: Use GetDiscriminatorValue for TPT and TPC in queries
- Make it default to `.ShortName()`
- Add validation that for non-TPH all values are unique strings

### PR DESCRIPTION
**Pull request by smitpatel**
Add model validation when values are not unique
Resolves #28054

### CODE REVIEW

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 1]
```diff
@@ -1391,6 +1391,24 @@ private static bool IsIdentifyingPrincipal(IEntityType dependentEntityType, IEnt
...
+                {
+                    if (!derivedType.ClrType.IsInstantiable())
+                    {
+                        continue;
+                    }
+                    var discriminatorValue = derivedType.ShortName();
+
+                    if (discriminatorValues.TryGetValue(discriminatorValue, out var duplicateEntityType))
+                    {
+                        throw new InvalidOperationException("TBD");
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1407]: Exception message?
Specifically I am not sure what to tell to users in how to resolve the issue.

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 2]
```diff
@@ -1391,6 +1391,24 @@ private static bool IsIdentifyingPrincipal(IEntityType dependentEntityType, IEnt
...
+
+                var derivedTypes = entityType.GetDerivedTypesInclusive().ToList();
+                var discriminatorValues = new Dictionary<object, IEntityType>();
+                foreach (var derivedType in derivedTypes)
+                {
+                    if (!derivedType.ClrType.IsInstantiable())
+                    {
+                        continue;
+                    }
+                    var discriminatorValue = derivedType.ShortName();
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1403]:
```suggestion
                    var discriminatorValue = derivedType.GetDiscriminatorValue();
```

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1407]: Use `EntityTypeBuilder.Metadata.SetDiscriminatorValue()`

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1403]: Also throw if it's not a `string`

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1407]: Can you suggest exception message? 😅

Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1407]: > The short name for '{entityType1}' is '{discriminatorValue}' which is the same for '{entityType2}'. Every concrete entity type in the hierarchy must have a unique short name. Either rename one of the types or call `modelBuilder.Entity<TEntity>().Metadata.SetDiscriminatorValue("NewShortName")`

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1403]: Exception message for this?

Comment (AndriySvyryd) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1403]: > The specified discriminator value '{value}' for '{entityType1}' is not a string. Configure a string discriminator value instead.

File: test/EFCore.Design.Tests/Scaffolding/Internal/CSharpRuntimeModelCodeGeneratorTest.cs [hunk 3]
```diff
@@ -1828,7 +1830,7 @@ public static void CreateAnnotations(RuntimeEntityType runtimeEntityType)
                     Assert.False(principalDerived.IsOwned());
                     Assert.IsType<ConstructorBinding>(principalDerived.ConstructorBinding);
                     Assert.Equal(ChangeTrackingStrategy.Snapshot, principalDerived.GetChangeTrackingStrategy());
-                    Assert.Null(principalDerived.GetDiscriminatorValue());
+                    Assert.Null(principalDerived[CoreAnnotationNames.DiscriminatorValue]);
```
Comment (AndriySvyryd) [inline test/EFCore.Design.Tests/Scaffolding/Internal/CSharpRuntimeModelCodeGeneratorTest.cs:1833]: Assert the new value, the runtime type doesn't store the discriminator value in an annotation

File: test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs [hunk 4]
```diff
@@ -180,9 +180,9 @@ private static void AssertDefaultMappings(IRelationalModel model, Mapping mappin
                 Assert.False(specialityColumn.IsNullable);
 
                 Assert.Null(customerType.FindDiscriminatorProperty());
-                Assert.Null(customerType.GetDiscriminatorValue());
+                Assert.Null(customerType[CoreAnnotationNames.DiscriminatorValue]);
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs:183]: Change these to assert the new value

File: test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs [hunk 5]
```diff
@@ -393,9 +393,9 @@ private static void AssertViews(IRelationalModel model, Mapping mapping)
...
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs:396]: Remove these

File: test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs [hunk 6]
```diff
@@ -765,9 +765,9 @@ private static void AssertTables(IRelationalModel model, Mapping mapping)
                 Assert.NotNull(anotherSpecialCustomerDbIndex.MappedIndexes.Single());
 
                 Assert.Null(customerType.FindDiscriminatorProperty());
-                Assert.Null(customerType.GetDiscriminatorValue());
+                Assert.Null(customerType[CoreAnnotationNames.DiscriminatorValue]);
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Tests/Metadata/RelationalModelTest.cs:768]: Remove these

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 7]
```diff
@@ -1391,6 +1391,30 @@ private static bool IsIdentifyingPrincipal(IEntityType dependentEntityType, IEnt
...
                 ValidateNonTphMapping(entityType, forTables: false);
                 ValidateNonTphMapping(entityType, forTables: true);
+
+                var derivedTypes = entityType.GetDerivedTypesInclusive().ToList();
+                var discriminatorValues = new Dictionary<string, IEntityType>();
+                foreach (var derivedType in derivedTypes)
+                {
+                    if (!derivedType.ClrType.IsInstantiable())
+                    {
+                        continue;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:1401]: Either change `GetDiscriminatorValue` to return `null` for non-instantiable types or validate that it's unique too

File: test/EFCore.Tests/ModelBuilding/InheritanceTestBase.cs [hunk 8]
```diff
@@ -80,8 +80,8 @@ public virtual void Can_specify_discriminator_values_first()
 
             var model = modelBuilder.FinalizeModel();
 
-            Assert.Null(model.FindEntityType(typeof(PBase)).GetDiscriminatorValue());
-            Assert.Null(model.FindEntityType(typeof(P)).GetDiscriminatorValue());
+            Assert.Null(model.FindEntityType(typeof(PBase))[CoreAnnotationNames.DiscriminatorValue]);
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilding/InheritanceTestBase.cs:83]: You can revert these

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**